packages/xstate-solid/src/deriveServiceState.ts

```diff
@@ -22,31 +22,40 @@ export const deriveServiceState = <
   StateSnapshot extends AnyState,
   StateReturnType = CheckSnapshot<StateSnapshot>
 >(
-  state: StateSnapshot | unknown
+  state: StateSnapshot | unknown,
+  prevState?: StateSnapshot | unknown
 ): StateReturnType => {
   if (isState(state)) {
-    return {
-      ...state,
-      toJSON: state.toJSON,
-      toStrings: state.toStrings,
-      can(...args: Parameters<StateSnapshot['can']>) {
-        // tslint:disable-next-line:no-unused-expression
-        this.value; // reads state.value to be tracked
-        // tslint:disable-next-line:no-unused-expression
-        this.context; // reads state.context to be tracked
-        return state.can(args[0]);
-      },
-      hasTag(...args: Parameters<StateSnapshot['hasTag']>) {
-        // tslint:disable-next-line:no-unused-expression
-        this.value; // reads state.value to be tracked
-        return state.hasTag(args[0]);
-      },
-      matches(...args: Parameters<StateSnapshot['matches']>) {
-        // tslint:disable-next-line:no-unused-expression
-        this.value; // reads state.value to be tracked
-        return state.matches(args[0] as never);
-      }
-    } as StateReturnType;
+    if (prevState && isState(prevState)) {
+      const {
+        toJSON,
+        toStrings,
+        can,
+        hasTag,
+        matches,
+        ...updatedState
+      } = state;
+      return {
+        ...updatedState,
+        toJSON,
+        toStrings,
+        can:
+          prevState.value !== state.value || prevState.context !== state.context
+            ? can
+            : prevState.can,
+        hasTag: prevState.value !== state.value ? hasTag : prevState.hasTag,
+        matches: prevState.value !== state.value ? matches : prevState.matches
```

Comment:
Hm, so the key here is to just reuse the functions when possible. So in a way - if `state.can()` depends on smth like, hm, `state.context.something.deep` then we won't reevaluate this only when this `.deep` changes but whenever the `state.context` changes.

Is is impossible to get a "fine-grained" reactivity for those methods?

Reply:
Probably not with can, we don't have any way of knowing a specific state only depends on a certain path in context.

Reply:
what if we'd actually create a special `.can` here (once per actor) and always call real `State['can']` with our proxy as its `this`? The `State['can']` wouldn't really know that it's being called with something different than usual - but since it would access proxy values, it would opt-into tracking them correctly. Or at least - this is what I imagine **maybe** being possible. It's quite fuzzy to me though if that's actually possible.

Reply:
I don't think Solid will be able to `.can`'s deep context dependencies with how `.can` works. The values being tracked are abstracted away from the method. On top of that, binding using our proxy state as `this` for `.can` will likely cause problems in the sub methods